cmd/consumer/main.go

```diff
@@ -0,0 +1,230 @@
+		return errors.Wrap(err, "creating logger")
+	}
+	defer logger.Sync()
+
+	go func() {
+		// run the consumer the first time when the goroutine starts
+		consumer(ctx, cfg, logger)
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-time.Tick(cfg.WorkerIntervalDuration):
+				logger.Info("running consumer")
+				consumer(ctx, cfg, logger)
+			}
+		}
+	}()
+
+	// handle interrupts and propagate the changes across the consumer pipeline
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt)
+	<-c
+	cancel()
+
+	return nil
+}
+
+// consumer fetches the top story ids from Hacker News and passes them along a pipeline to be fetched and inserted
+func consumer(ctx context.Context, cfg *Config, logger *zap.Logger) error {
+	logger.Info("opening database connection")
+	store := postgres.NewItemStore()
+	if err := store.Open(ctx, cfg.DatabaseDSN); err != nil {
+		return errors.Wrap(err, "opening store db connection")
+	}
+	defer store.Close(ctx)
+
+	logger.Info("creating HN client")
+	client := hn.NewClient()
+
+	ids, err := client.FetchTopStories()
+	if err != nil {
+		return errors.Wrap(err, "fetching top stories")
+	}
+	logger.Info("fetched top story ids", zap.Int("count", len(ids)))
+
+	// convert the top story ids into a channel of ids
+	idStream := generator(ctx, ids)
+
+	// create workers to fetch items from the HN API
+	fetchers := make([]<-chan fetchItemResult, cfg.WorkerCount)
+	for i := 0; i < cfg.WorkerCount; i++ {
+		fetchers[i] = fetchItem(ctx, idStream, client)
+	}
+
+	// fan-in the fetched items and decide what to do with them
+	for result := range fanInFetchedItems(ctx, fetchers...) {
+		if result.Error != nil {
+			logger.Error("error fetching item", zap.Error(result.Error))
+			continue
+		}
+
+		if result.Item.Dead || result.Item.Deleted {
+			// ignore dead or deleted items
+			break
+		}
+
+		logger.Info("inserting item", zap.Int("id", result.Item.ID))
+		store.Insert(ctx, result.Item)
+	}
+
+	logger.Info("finished inserting items")
+
+	return nil
+}
+
+func generator(ctx context.Context, ids []int) <-chan int {
+	idStream := make(chan int)
+
+	go func() {
+		defer close(idStream)
+
+		for _, id := range ids {
+			select {
+			case <-ctx.Done():
+				return
+			case idStream <- id:
+			}
+		}
+	}()
+
+	return idStream
+}
+
+type fetchItemResult struct {
+	Item  *hn.Item
+	Error error
+}
+
+func fetchItem(ctx context.Context, idStream <-chan int, client *hn.Client) <-chan fetchItemResult {
+	itemStream := make(chan fetchItemResult)
+
+	go func() {
+		defer close(itemStream)
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case id, ok := <-idStream:
+				if !ok {
+					return
+				}
+
+				item, err := client.FetchItem(id)
+				// wrap both the return item and the error so that whatever consumes this channel can deal with the errors more appropriately
+				result := fetchItemResult{Item: item, Error: err}
+
+				select {
+				case <-ctx.Done():
+					return
+				case itemStream <- result:
+				}
```

Comment:
Unsure if you need this second select here.

you should be able to do a raw `itemStream <- result`

Can you give it a try and let me know what happens, will be interesting to see.

Reply:
Just tried this and the program hangs on interrupt if the itemStream isn't being read from.